### DISCUSSION TITLE
Prefix MQTT topics with username

### DISCUSSION THREAD

**Discussion by Superbert25** [arendst/Tasmota#13778] [Q&A] [answered]
Hi,
I'm looking for a way to prefix the MQTT topic with the value of the MQTT user field. Just something like %06x that inserts the MAC, but then for user.

The intention is to combine this with the ACLs in mosquitto (using %u). So, I want to restrict subscribing and publishing topics based upon "MQTT user".

I did manage to create a custom build with TLS enabled, so making small changes in the config / code is viable to me.
Thanks for your help.

**Comment by sfromis** [comment]
You need no code changes. On the "Configure MQTT" page, you can already customize your MQTT topics, including not having the "prefix" as the first part. Same options are available build time, if you prefer.

**Comment by Superbert25** [comment]
First of all, thanks for your feedback.

I think my question was not clear enough, so let me try again:

The user gets his authentication  credentials (from registration) and puts them in the MQTT config of the Tasmota device. 
Next, the Tasmota device should use the username as topic prefix (automatically). 

This prefix is used in Mosquitto to determine the ACL (%u variable). This way I can separate multiple users in the broker.

**Reply by sfromis** [reply]
The user can put the user name into the Tasmota config, you can also have some onboarding procedure to automate that. Details depend on your undefined workflow around "gets his authentication credentials".

**Comment by barbudor** [comment, marked as answer]
On my side, I'm using the same name for topic and user, each device having a unique topic/user. So I can use %u in mosquitto ACL.

You can also use full topic such as '<user>/%topic%/%prefix%' so your ACL can start with %u.
But it's your provisioning mecanism that will have to set the full topic this way depending on the user. There is no special variable like %user%.
Considering that you have anyhow to configure MqttUser, configuring FullTopic accordingly shouldn't be a problem

**Reply by Superbert25** [reply]
IMHO It’s a pitty there’s no variable for “user” as it’s a possible point of failure for the end-user when he has to duplicate information manually.

Thanks for clearing up.

R. Eeg

**Reply by sfromis** [reply]
In general, you can automate the duplication, like by using rules.

**Reply by Superbert25** [reply]
Interesting! Can you point me in the right direction? Is there a “user” variable available in the rules?

**Reply by sfromis** [reply]
No user variable, again.

What you can do is to use [rules](https://tasmota.github.io/docs/Rules/) (some reading needed) to grab the output of one command (including what's used by the web UI, watch the console closely), put it into a variable, and issue new command(s) including the use of that value. If only one extra command, you do not need to store it in a variable.

**Reply by Superbert25** [reply]
I Tested the following (and it works as expected):

`rule1 on mqttuser#data do backlog Topic %value% endon`

This rule puts the same value in Topic as what has been input in mqttuser.

Is there a way to concatenate "%06X" to the Topic by rule? I have tried by using a "+", but the "+" is seen as a character. I couldn't find anything on concatenating / adding something to a string in the docs.

Cheers,

Eeg.

**Reply by Superbert25** [reply]
Solved...it was easier than I ever expected, I used the following:

`rule1 on mqttuser#data do backlog Topic %value%/%06X endon`

**Reply by barbudor** [reply]
Good trick.
Hapy that you found a way through
👍
